### PR TITLE
[server] Sync token refresh

### PR DESCRIPTION
**Pull request by geropl**
## Description
Attempt two to fix token refresh (attempt one is [here](https://github.com/gitpod-io/gitpod/pull/19452)) using a redis mutex.

Code is guarded with the `sync_refresh_token_exchange` feature flag.

## Related Issue(s)

Fixes EXP-1413

## How to test
 - start a workspace on a repo on gitlab.com :heavy_check_mark: 
 - extended test:
   - invalidate your GitLab token (by changing the expiryDate in the DB)
   - start another workspace on GitLab
   - see how the token got refreshed :heavy_check_mark: 

#### Preview status

gitpod:summary

## Build Options

<details>
<summary>Build</summary>

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer</summary>

- [ ] analytics=segment
- [ ] with-dedicated-emulation
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment / Integration Tests</summary>

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [ ] /werft with-preview
- [ ] /werft with-large-vm
- [x] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [x] /werft preemptible
      Saves cost. Untick this only if you're really sure you need a non-preemtible machine.
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`. If enabled, `with-preview` and `with-large-vm` will be enabled.
- [ ] with-monitoring
</details>

/hold

### CODE REVIEW

**Review by mustard-mh — APPROVED** [review]
Code LGTM, but I don't understand why we add endpoint (Caddyfile) for `public.v1` APIs. Not block but would be good to understand

**Comment by mustard-mh** [comment]
Sorry, commented on wrong PR...

File: components/server/src/user/token-service.ts [hunk 1]
```diff
@@ -42,6 +55,54 @@ export class TokenService implements TokenProvider {
...
+                [`token-refresh-${host}-${userId}`],
+                2000, // After 2s without extension the lock is released
+                doRefreshToken,
+                { retryCount: 10, retryDelay: 500 }, // We wait at most 10s until we give up, and conclude that we can't refresh the token now.
+            );
+            return refreshedToken;
+        } catch (err) {
+            if (RedisMutex.isLockedError(err)) {
+                log.error({ userId }, `Failed to refresh token (timeout waiting on lock)`, err, { host });
+                throw new Error(`Failed to refresh token (timeout waiting on lock)`);
```
Comment (mustard-mh) [inline components/server/src/user/token-service.ts:100]: Use ApplicationError?

File: components/server/src/user/token-service.ts [hunk 2]
```diff
@@ -42,6 +55,54 @@ export class TokenService implements TokenProvider {
...
+
+        const doRefreshToken = async () => {
+            // Check: Current token so we can actually refresh?
+            const token = await this.userDB.findTokenForIdentity(identity);
+            if (!token) {
+                return undefined;
+            }
+
+            const aboutToExpireTime = new Date();
+            aboutToExpireTime.setTime(aboutToExpireTime.getTime() + 5 * 60 * 1000);
```
Comment (mustard-mh) [inline components/server/src/user/token-service.ts:74]: [nit] will a provider's token always expired in 5 minutes, then never make it works? 🤣

File: components/server/src/user/token-service.ts [hunk 3]
```diff
@@ -29,6 +31,17 @@ export class TokenService implements TokenProvider {
 
     async getTokenForHost(user: User | string, host: string): Promise<Token | undefined> {
         const userId = User.is(user) ? user.id : user;
+
+        // EXPERIMENT(sync_refresh_token_exchange)
+        const syncRefreshTokenExchange = await getExperimentsClientForBackend().getValueAsync(
+            "sync_refresh_token_exchange",
```
Comment (mustard-mh) [inline components/server/src/user/token-service.ts:37]: retry_refresh_token_exchange and sync_refresh_token_exchange are different

- sync_refresh_token_exchange will store them in database
- retry_refresh_token_exchange will revoke (internally) token after it's used

Is it intent?

File: components/server/src/user/token-service.ts [hunk 4]
```diff
@@ -29,6 +31,17 @@ export class TokenService implements TokenProvider {
...
+            false,
+            {},
```
Comment (mustard-mh) [inline components/server/src/user/token-service.ts:39]: Should we add host as a property of Feature Flag? Just in case they have multiple SCM

**Comment by mustard-mh** [comment]
Approved to unblock

File: components/server/src/user/token-service.ts (hunk 1 above)
Comment (geropl) [reply components/server/src/user/token-service.ts:100]: I was not sure where it gets handled, so I kept is as before (no ApplkicationError), just changed message.

File: components/server/src/user/token-service.ts (hunk 3 above)
Comment (geropl) [reply components/server/src/user/token-service.ts:37]: The intention behind both paths is the same. Deep down they both call:
```
await authProvider.refreshToken(user); // updating/inserting a fresh token into the DB as side effect
return await this.userDB.findTokenForIdentity(identity);   // return said token
```

It's just that the existing path has this poor-man's sync approach (from before we had redis), which I did not want for the pure redis solution.
What I basically did for `sync_refresh_token_exchange` is to wrap it into a redis mutex, nothing more. The rest is noise/slight code re-arrangements.

File: components/server/src/user/token-service.ts (hunk 4 above)
Comment (geropl) [reply components/server/src/user/token-service.ts:39]: We are interested to role this out for all, so I think we should test it as such. The difference between SCMs should be properly abstracted behind the Token abstraction.

**Pull request merged**